pkg/deploy/deploy.go

```diff
@@ -0,0 +1,260 @@
+package deploy
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"time"
+
+	azcompute "github.com/Azure/azure-sdk-for-go/services/compute/mgmt/2019-03-01/compute"
+	azresources "github.com/Azure/azure-sdk-for-go/services/resources/mgmt/2018-05-01/resources"
+	"github.com/Azure/go-autorest/autorest"
+	"github.com/Azure/go-autorest/autorest/to"
+	"github.com/sirupsen/logrus"
+
+	"github.com/Azure/ARO-RP/pkg/deploy/generator"
+	"github.com/Azure/ARO-RP/pkg/util/arm"
+	"github.com/Azure/ARO-RP/pkg/util/azureclient/mgmt/compute"
+	"github.com/Azure/ARO-RP/pkg/util/azureclient/mgmt/network"
+	"github.com/Azure/ARO-RP/pkg/util/azureclient/mgmt/resources"
+	"github.com/Azure/ARO-RP/pkg/util/ready"
+)
+
+var _ Deployer = (*deployer)(nil)
+
+type Deployer interface {
+	PreDeploy(context.Context, *logrus.Entry) (string, error)
+	Deploy(context.Context, *logrus.Entry, string) error
+	Upgrade(context.Context, *logrus.Entry) error
+}
+
+type deployer struct {
+	log         *logrus.Entry
+	deployments resources.DeploymentsClient
+	groups      resources.GroupsClient
+	vmss        compute.VirtualMachineScaleSetsClient
+	vmssvms     compute.VirtualMachineScaleSetVMsClient
+	network     network.PublicIPAddressesClient
+
+	cli *http.Client
+
+	parameters        *arm.Parameters
+	version           string
+	resourceGroupName string
+	subscriptionID    string
+	location          string
+}
+
+// New initiates new deploy utility object
+func New(ctx context.Context, log *logrus.Entry, authorizer autorest.Authorizer, gitCommit string) (*deployer, error) {
+	d := &deployer{
+		log:               log,
+		version:           gitCommit,
+		resourceGroupName: os.Getenv("AZURE_RP_RESOURCEGROUP_NAME"),
+		subscriptionID:    os.Getenv("AZURE_SUBSCRIPTION_ID"),
+		location:          os.Getenv("LOCATION"),
+	}
+
+	d.cli = &http.Client{
+		Timeout: 5 * time.Second,
+	}
+
+	d.deployments = resources.NewDeploymentsClient(d.subscriptionID, authorizer)
+	d.groups = resources.NewGroupsClient(d.subscriptionID, authorizer)
+	d.vmss = compute.NewVirtualMachineScaleSetsClient(d.subscriptionID, authorizer)
+	d.vmssvms = compute.NewVirtualMachineScaleSetVMsClient(d.subscriptionID, authorizer)
+	d.network = network.NewPublicIPAddressesClient(d.subscriptionID, authorizer)
+
+	return d, nil
```

Comment:
Is the idea to eventually check for and possibly return an error here? Right now it appears as though you could just return a `*deployer` that's assembled in one step and remove the error return.